network-mux/test/Test/Mux.hs

```diff
@@ -1219,9 +1222,62 @@ dummyAppToChannel DummyApp {daAction, daRunTime} = \_ -> do
          DummyAppSucceed -> return ((), Nothing)
          DummyAppFail    -> throwM $ MuxError (MuxShutdown Nothing) "App Fail"
 
+data DummyRestartingApps =
+    DummyRestartingResponderApps [(DummyApp, Int)]
+  | DummyRestartingInitiatorApps [(DummyApp, Int)]
+  | DummyRestartingInitiatorResponderApps [(DummyApp, Int)]
+  deriving Show
+
+instance Arbitrary DummyRestartingApps where
+    arbitrary = do
+        nums <- listOf1 $ arbitrary
+        apps <- mapM genApp $ nub nums
+        mode <- arbitrary
+        case mode of
+             InitiatorMode          -> return $ DummyRestartingInitiatorApps apps
+             ResponderMode          -> return $ DummyRestartingResponderApps apps
+             InitiatorResponderMode -> return $ DummyRestartingInitiatorResponderApps apps
+      where
+        genApp num = do
+            app <- DummyApp num DummyAppSucceed <$> arbitrary <*> arbitrary
+            restarts <- choose (0, 5)
+            return (app, restarts)
+
+
+dummyRestartingAppToChannel :: forall a m.
+                     ( MonadAsync m
+                     , MonadCatch m
+                     , MonadTimer m
+                     )
+                  => (DummyApp, a)
+                  -> (Channel m -> m ((DummyApp, a), Maybe BL.ByteString))
+dummyRestartingAppToChannel (app, r) = \_ -> do
+    threadDelay $ daRunTime app
+    case daAction app of
+         DummyAppSucceed -> return ((app, r), Nothing)
+         DummyAppFail    -> throwM $ MuxError (MuxShutdown Nothing) "App Fail"
+
+
 appToInfo :: MiniProtocolDirection mode -> DummyApp -> MiniProtocolInfo mode
 appToInfo d da = MiniProtocolInfo (daNum da) d defaultMiniProtocolLimits
 
+triggerApp :: forall m.
+              ( MonadAsync m
+              , MonadSay m
+              , MonadTime m
+              , MonadTimer m
+              )
+            => MuxBearer m
+            -> DummyApp
+            -> m ()
+triggerApp bearer app = do
+    let chan = muxBearerAsChannel bearer (daNum app) InitiatorDir
+    traceWith verboseTracer $ "app waiting " ++ (show $ daNum app)
+    threadDelay (daStartAfter app)
+    traceWith verboseTracer $ "app starting " ++ (show $ daNum app)
+    send chan $ BL.singleton 0xa5
```

Comment:
why just a single byte, not the whole conversation?

Reply:
I see, it is not using `Test.Mux.ReqResp` protocol, which I hoped for.  When we are using `dummyAppToChannel` we never ready any bytes from mux ingress queue.  If we use a proper protocol the tests would be more comprehensive.

Another testing strategy that could be useful is using `IOSim` and tracing to ensure that the mux is in the right state on restarts. We do that for other components: block-fetch, connection-manager, peer-to-peer governor.

Reply:
See https://github.com/input-output-hk/ouroboros-network/pull/2629#discussion_r500214138 . ReqResp byte tests are covered in other testcases.